jerry-core/ecma/builtin-objects/ecma-builtin-json.c

```diff
@@ -1260,13 +1260,33 @@ ecma_builtin_json_serialize_property (ecma_json_stringify_context_t *context_p,
   /* 11. */
   if (ecma_is_value_object (value) && !ecma_op_is_callable (value))
   {
+    ecma_value_t is_array = ecma_is_value_array (value);
+
+#if ENABLED (JERRY_ES2015)
+    if (ECMA_IS_VALUE_ERROR (is_array))
+    {
+      return is_array;
+    }
+#endif /* ENABLED (JERRY_ES2015) */
+
     ecma_object_t *obj_p = ecma_get_object_from_value (value);
 
     ecma_value_t ret_value;
+
     /* 10.a */
-    if (ecma_get_object_type (obj_p) == ECMA_OBJECT_TYPE_ARRAY)
+    if (ecma_is_value_true (is_array))
     {
-      ret_value = ecma_builtin_json_serialize_array (context_p, obj_p);
+      ecma_object_t *array_obj_p = obj_p;
+
+#if ENABLED (JERRY_ES2015)
+      if (ECMA_OBJECT_IS_PROXY (obj_p))
```

Comment:
You should pass `obj_p` to json_serialize array.

The assertion in the beginning of `json_serialize array` should be changed to  `ecma_is_value_array `.

The 6. step of this function should be changed to `Get(O, "length")` whenever the `O` is proxy, otherwise it should remain the same code as before.